src/object/cli_obj.c

```diff
@@ -4760,9 +4760,17 @@ obj_comp_cb(tse_task_t *task, void *data)
 						obj_auxi->tx_uncertain = 1;
 					else
 						obj_auxi->nvme_io_err = 1;
-				} else if (task->dt_result != -DER_NVME_IO) {
-					/* Don't retry update for CSUM & UNCERTAIN errors */
-					obj_auxi->io_retry = 0;
+				} else {
+					if (task->dt_result == -DER_CSUM) {
+						/** Retry once on checksum error on update */
+						if (!obj_auxi->csum_retry)
```

Comment:
it possibly will cause always retry if it always return -DER_CSUM, because above L4689 clears the "obj_auxi->csum_retry" flag ahead.
So a simple fix may could check "if(obj_auxi->retry_cnt == 0)" here, it will not retry if previously retried by TIMEOUT, but looks fine.
Another option is change L4689 to don't clear the flag for UPDATE operation.

Current client code for retry handling is not clean nor easy to understand (many cases added by many persons), need to be refined/cleanup later.

Reply:
Good catch. From our testing, it sounds like we can have checksum errors a few times, so retrying several times (indefinitely?) sounds like a reasonable approach for now.

Reply:
If with a limited MAX number of retrying, or with a limited time window retrying that seems better than indefinitely retrying forever as that possibly cause dead loop.